### ISSUE TITLE
Is it possible to construct a VCache or VSpace mock value in a pure way for testing purposes?

### ISSUE THREAD

**Issue by Wizek** [dmbarbour/haskell-vcache#7]
Mainly asking for https://github.com/dmbarbour/vcache-trie/issues/2, but could be nice for other data types too.

**Comment by Ericson2314** [comment]
I made the interface a type class in https://github.com/Ericson2314/haskell-vcache. That could help.

**Comment by Wizek** [comment]
@Ericson2314 Hmm, interesting, thanks for letting me know! It's [here](https://github.com/dmbarbour/haskell-vcache/commit/8703fb722ac171a52861b7ce5c5ad5c5581b2d56#diff-532e513c07c2a85eb6b078910d4bc0e4R42), right? This may help.

On an unrelated note, I wonder, are the two of us the only ones who use/have used vcache? :)
I wonder how come the adoption is as low as it currently is, i.e. in contrast to acid-state.

**Comment by verement** [comment]
I use vcache for [EtaMOO](https://github.com/verement/etamoo), so there are at least three of us.

**Comment by Ericson2314** [comment]
Yeah, that and the previous commit.

Unfortunately I never got around to actually using VCache for anything. (My hope was to make an alternative backend with vrefs content-addressed to work with something like IPFS, hence the type classes, and needless to say I never finished the project.) It's a shame though, because I did and still think the concept is super good.

**Comment by Wizek** [comment]
@Ericson2314 Super good indeed. I do use it in an app, almost daily, and I quite enjoy it's effortless and transparent persistence. :)

Also, never say never; maybe just haven't yet. ;)

**Comment by Ericson2314** [comment]
@Wizek well, I think it's important to be honest to oneself and others about stalled hobby projects. Let the presumption be I'll never return to it, and maybe we'll be pleasantly surprised :).

**Comment by dhruvio** [comment]
Our team is considering using `vcache` for the persistence layer of various web applications. The goal is to use Haskell types to define "schema", and have them be compatible with our application layer (i.e. a process vending data over a JSON API). We're not sure how feasible this would be given the high concurrency demands of web apps (LMDB seems well-suited to this task), and the need to eventually scale the persistence layer independently of the application layer depending on process usage. It seems like there would a fair amount of work involved to make this achievable, but the payoff of having type safety from persistence to API could make it worthwhile.

**Comment by dmbarbour** [comment]
Please take over the project and maintain it yourself if you want to use it.

I mostly abandoned it because Haskell's performance wasn't good enough for my purposes. Not performance of vcache but in other areas - plugins, runtime compilation.

**Comment by Wizek** [comment]
Yeah, I too have mostly abandoned Haskell, partly for performance, and partly for other reasons too. And I miss VCache-like database approaches in other ecosystems.

**Comment by dhruvio** [comment]
@dmbarbour Thanks for the update. We are still reviewing various approaches, but I will reach out to you if we decide to use `VCache`.
